### ISSUE TITLE
ModuleNotFoundError: No module named 'gtdbtk'

### ISSUE THREAD

**Issue by franciscozorrilla** [Ecogenomics/GTDBTk#241]
Hello,

tldr: gtdbtk works on login node but fails when submitting jobs to the cluster.

I followed the bioconda [installation instructions](https://ecogenomics.github.io/GTDBTk/running/bioconda.html):

```
conda create -n gtdbtk -c bioconda gtdbtk
download-db.sh
```
Database path variable and dependencies look good:

```
$ paste /g/scb2/patil/zorrilla/conda/envs/gtdbtk/etc/conda/activate.d/gtdbtk.sh
export GTDBTK_DATA_PATH=/g/scb2/patil/zorrilla/conda/envs/gtdbtk/share/gtdbtk-1.1.0/db/

$ conda list
# packages in environment at /g/scb2/patil/zorrilla/conda/envs/gtdbtk:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                      1_llvm    conda-forge
boost                     1.70.0           py38h9de70de_1    conda-forge
boost-cpp                 1.70.0               h8e57a91_2    conda-forge
bzip2                     1.0.8                h516909a_2    conda-forge
ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
capnproto                 0.6.1                hfc679d8_1    conda-forge
certifi                   2020.4.5.1       py38h32f6830_0    conda-forge
dendropy                  4.4.0                      py_1    bioconda
fastani                   1.3                  he1c1bb9_0    bioconda
fasttree                  2.1.10               h14c3975_3    bioconda
gsl                       2.5                  h294904e_1    conda-forge
gtdbtk                    1.1.0                      py_0    bioconda
hmmer                     3.1b2                         3    bioconda
icu                       64.2                 he1b5a44_1    conda-forge
ld_impl_linux-64          2.34                 h53a641e_0    conda-forge
libblas                   3.8.0               16_openblas    conda-forge
libcblas                  3.8.0               16_openblas    conda-forge
libffi                    3.2.1             he1b5a44_1007    conda-forge
libgcc-ng                 9.2.0                h24d8f2e_2    conda-forge
libgfortran-ng            7.3.0                hdf63c60_5    conda-forge
liblapack                 3.8.0               16_openblas    conda-forge
libopenblas               0.3.9                h5ec1e0e_0    conda-forge
libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
llvm-openmp               10.0.0               hc9558a2_0    conda-forge
mash                      2.2.2                h3d38be6_1    bioconda
ncurses                   6.1               hf484d3e_1002    conda-forge
numpy                     1.18.1           py38h8854b6b_1    conda-forge
openssl                   1.1.1f               h516909a_0    conda-forge
pip                       20.0.2                     py_2    conda-forge
pplacer                   1.1.alpha19                   1    bioconda
prodigal                  2.6.3                h516909a_2    bioconda
python                    3.8.2           h8356626_5_cpython    conda-forge
python_abi                3.8                      1_cp38    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
setuptools                46.1.3           py38h32f6830_0    conda-forge
sqlite                    3.30.1               hcee41ef_0    conda-forge
tk                        8.6.10               hed695b0_0    conda-forge
wheel                     0.34.2                     py_1    conda-forge
xz                        5.2.5                h516909a_0    conda-forge
zlib                      1.2.11            h516909a_1006    conda-forge
```

Then I ran the test command: 
```
$ gtdbtk test --out_dir testGTDBtk --cpus 3
[2020-04-17 17:26:14] INFO: GTDB-Tk v1.1.0
[2020-04-17 17:26:14] INFO: gtdbtk test --out_dir testGTDBtk --cpus 3
[2020-04-17 17:26:14] INFO: Using GTDB-Tk reference data version r89: /g/scb2/patil/zorrilla/conda/envs/gtdbtk/share/gtdbtk-1.1.0/db/
[2020-04-17 17:26:14] INFO: Command: gtdbtk classify_wf --genome_dir testGTDBtk/genomes --out_dir testGTDBtk/output --cpus 3
[2020-04-17 17:27:18] ERROR: The test returned a non-zero exit code.
[2020-04-17 17:27:18] ERROR: A detailed summary of the execution log can be found here: testGTDBtk/test_execution.log
[2020-04-17 17:27:18] ERROR: The test has failed.

[2020-04-17 17:27:18] ERROR: Controlled exit resulting from early termination.

```
Although the test failed, likely due to lack of RAM since I was running on the login node, it seems like things are installed properly. 

However, when I try to submit GTDB-Tk jobs to the cluster (with the suggested 100GB RAM) I get this error message:

```
$ paste logs/.bx02-06.testGTDBtk.out.log
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	testGTDBtk
	1

[Fri Apr 17 17:02:24 2020]
rule testGTDBtk:
    input: /home/zorrilla/workspace/tara/dna_bins
    output: /home/zorrilla/workspace/tara/GTDBtk
    jobid: 0
    benchmark: /home/zorrilla/workspace/tara/GTDBtk.benchmark.txt

Traceback (most recent call last):
  File "/g/scb2/patil/zorrilla/conda/envs/gtdbtk/bin/gtdbtk", line 36, in <module>
    from gtdbtk import version
ModuleNotFoundError: No module named 'gtdbtk'
[Fri Apr 17 17:02:32 2020]
Error in rule testGTDBtk:
    jobid: 0
    output: /home/zorrilla/workspace/tara/GTDBtk
    shell:
        
        module load Anaconda3
        set +u;source activate gtdbtk;set -u;
        cd $SCRATCHDIR
        cp -r /home/zorrilla/workspace/tara/dna_bins .

        gtdbtk classify_wf --genome_dir dna_bins --out_dir GTDBtk -x fa --cpus 8
        mv GTDBtk $(dirname /home/zorrilla/workspace/tara/GTDBtk)

        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message

```

I am not really sure if the conda environment is not being loaded properly withing the job, or if the installation was not completed successfully, or perhaps something else is going wrong. Any suggestions?

Best,
Francisco

**Comment by aaronmussig** [comment]
Hi Francisco,

I'm not sure what is causing the gtdbtk module not being found. I've not used snakemake often enough to know, but perhaps there is an issue with snakemake activating conda environments in the shell script?

If you can't find any info on that, there are a couple of things you can try which come to mind:
1. Verify that the python interpreter is correct by capturing the output of `which python` in the cluster job.
2. Try force the conda environment to be recognised by manually setting `PATH` and `PYTHONPATH`.
3. Try using a conda environment which uses `__main__` and specifies an entrypoint instead of using `bin/gtdbtk` via: 
    * `conda create -n gtdbtk-tmp -c aaronmussig gtdbtk-dev-241`
    * Note that you will need to update the GTDBTK_DATA_PATH variable in `envs/gtdbtk-tmp/etc/conda/activate.d/gtdbtk.sh` if you don't want to download the reference data again.

**Comment by susheelbhanu** [comment]
I have the same "no module named gtdbtk" error after installing via conda. I am not running it via a Snakefile. It's being run on a HPC though interactively for the moment. Here's the error message I'm getting:
```
Traceback (most recent call last):
  File "/home/users/sbusi/apps/miniconda3/envs/gtdbtk/bin/gtdbtk", line 36, in <module>
    from gtdbtk import version
ModuleNotFoundError: No module named 'gtdbtk'
```

**Comment by franciscozorrilla** [comment]
Hi @aaronmussig, thank you for the suggestions.
It seems to be working now after creating a new environment as you suggested:
```
$ tail -n 1000 -F logs/.bx08-10.testGTDBtk.out.log
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 16
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	testGTDBtk
	1

[Wed Apr 22 19:25:35 2020]
rule testGTDBtk:
    input: /home/zorrilla/workspace/tara/dna_bins
    output: /home/zorrilla/workspace/tara/GTDBtk
    jobid: 0
    benchmark: /home/zorrilla/workspace/tara/GTDBtk.benchmark.txt

/g/easybuild/x86_64/CentOS/7/haswell/software/Anaconda3/2019.07/bin/python
[2020-04-22 19:25:56] INFO: GTDB-Tk v1.1.0
[2020-04-22 19:25:56] INFO: gtdbtk classify_wf --genome_dir dna_bins --out_dir GTDBtk -x fa --cpus 16
[2020-04-22 19:25:56] INFO: Using GTDB-Tk reference data version r89: /g/scb2/patil/zorrilla/conda/envs/gtdbtk/share/gtdbtk-1.1.0/db/
[2020-04-22 19:25:56] INFO: Identifying markers in 46 genomes with 16 threads.
[2020-04-22 19:25:57] INFO: Running Prodigal V2.6.3 to identify genes.
==> Finished processing 46 of 46 (100.0%) genomes.
[2020-04-22 19:27:14] INFO: Identifying TIGRFAM protein families.
==> Finished processing 46 of 46 (100.0%) genomes.
[2020-04-22 19:27:44] INFO: Identifying Pfam protein families.
==> Finished processing 46 of 46 (100.0%) genomes.
[2020-04-22 19:27:47] INFO: Annotations done using HMMER 3.1b2 (February 2015)
[2020-04-22 19:27:48] INFO: Done.
[2020-04-22 19:27:56] INFO: Aligning markers in 46 genomes with 16 threads.
[2020-04-22 19:27:56] INFO: Processing 45 genomes identified as bacterial.
[2020-04-22 19:28:06] INFO: Read concatenated alignment for 23458 GTDB genomes.
==> Finished aligning 45 of 45 (100.0%) genomes.
[2020-04-22 19:28:32] INFO: Masking columns of multiple sequence alignment using canonical mask.
[2020-04-22 19:29:07] INFO: Masked alignment from 41155 to 5040 AAs.
[2020-04-22 19:29:07] INFO: 0 user genomes have amino acids in <10.0% of columns in filtered MSA.
[2020-04-22 19:29:07] INFO: Creating concatenated alignment for 23503 GTDB and user genomes.
[2020-04-22 19:29:07] INFO: Creating concatenated alignment for 45 user genomes.
[2020-04-22 19:29:07] INFO: Processing 1 genomes identified as archaeal.
[2020-04-22 19:29:08] INFO: Read concatenated alignment for 1248 GTDB genomes.
==> Finished aligning 1 of 1 (100.0%) genomes.
[2020-04-22 19:29:12] INFO: Masking columns of multiple sequence alignment using canonical mask.
[2020-04-22 19:29:13] INFO: Masked alignment from 32675 to 5124 AAs.
[2020-04-22 19:29:13] INFO: 0 user genomes have amino acids in <10.0% of columns in filtered MSA.
[2020-04-22 19:29:13] INFO: Creating concatenated alignment for 1249 GTDB and user genomes.
[2020-04-22 19:29:13] INFO: Creating concatenated alignment for 1 user genomes.
[2020-04-22 19:29:13] INFO: Done.
[2020-04-22 19:29:14] INFO: Placing 1 archaeal genomes into reference tree with pplacer using 16 cpus (be patient).
Placing genomes |##################################################| 1/1 (100.00%)
[2020-04-22 19:29:38] INFO: pplacer version: v1.1.alpha19-0-g807f6f3
[2020-04-22 19:29:39] INFO: Calculating average nucleotide identity using FastANI.
[2020-04-22 19:29:39] INFO: fastANI version: 1.3
==> Processing 36 of 36 (100.0%) comparisons.
[2020-04-22 19:29:41] INFO: 1 genome(s) have been classified using FastANI and pplacer.
[2020-04-22 19:29:41] INFO: Calculating RED values based on reference tree.
[2020-04-22 19:29:41] INFO: Placing 45 bacterial genomes into reference tree with pplacer using 16 cpus (be patient).
```
I also tried creating another environment and installing with pip, and that also seems to work. I suspect there is probably something wrong with the conda recipe considering @susheelbhanu's comment.

Thanks,
Francisco

**Comment by susheelbhanu** [comment]
Hey @aaronmussig I was able to install via the 'developer' version you provided above. However, I am now running into this issue:

```
[2020-04-26 13:53:26] ESC[1mINFO:ESC[0m Annotations done using HMMER 3.1b2 (February 2015)
[2020-04-26 13:53:27] ESC[1mINFO:ESC[0m Done.
[2020-04-26 13:53:27] ESC[1;31mERROR:ESC[0m ['bwa_sr_metaspades_hybrid.metabat.26', 'bwa_sr_metaspades_hybrid.metabat.25', 'bwa_sr_metaspades_hybrid.metabat.83_sub', 'bwa_sr_metaspades_hybrid.metabat.44_sub', 'bwa_sr_metaspades_hybrid.metabat.73', 'bwa_sr_metaspades_hybrid.metabat.46', 'bwa_sr_metaspades_hybrid.metabat.63_sub', 'bwa_sr_metaspades_hybrid.metabat.39', 'bwa_sr_metaspades_hybrid.metabat.55', 'bwa_sr_metaspades_hybrid.metabat.49_sub', 'bwa_sr_metaspades_hybrid.metabat.59'] are not present in the input list of genome to process.

[2020-04-26 13:53:27] ESC[1;31mERROR:ESC[0m Controlled exit resulting from an unrecoverable error or warning.

================================================================================
EXCEPTION: InconsistentGenomeBatch
  MESSAGE: Number of processed genomes in the identify directory is unequal to the list of input genomes.
```
Due to maxbin, I have some bins that have a "_sub" suffix, while others don't. When I look in the "identify" folder, some of these have been renamed with "_sub" being assigned to another bin. Any ideas on how to proceed beyond this?

Thank you!

**Comment by aaronmussig** [comment]
Hi @susheelbhanu and @franciscozorrilla

Thanks for reporting back, I've staged a change for the next release which will resolve the ModuleNotFoundEror on clusters.

Regarding the InconsistentGenomeBatch error is indicating that some genomes didn't generate output files under the identify/intermediate_results/ directory. A workaround might be to run each of the steps individually, but that is sidestepping the issue.

What version are you running?
What is the quality of those bins?
Can you provide the bins which were unsuccessful?

**Comment by susheelbhanu** [comment]
Thanks @aaronmussig! Here's some information you wanted:
```
gtdbtk --version
gtdbtk: version 1.1.0 Copyright 2017 Pierre Chaumeil
```
I re-ran the analyses and don't have the 'troubled' bins anymore. It went through without any issues this time. I can tell you though that last I checked with CheckM, some of them had around 15% contamination. 

Thanks again for your help!
